docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1115 @@
+
+In automatic mode, one simply runs the command `git bisect run test.sh` on some
+shell script you've written `test.sh`, and `git-bisect` performs the entire
+bisection search automatically. The shell script needs to exit with a return
+value that tells `git bisect` whether the revision it's run on is good (`exit 0`),
+bad (`exit 1`), or should be skipped (`exit 125`).
+
+How the test-script determines the presence of a regression is up to you: a
+typical approach is to measure against a baseline (eg. using
+`utils/process-stats-dir.py --compare-to-csv-baseline`, if your regression range
+covers compilers that all support `--output-stats-dir`). Alternatively, just
+measure raw time or instruction counts. An example script that uses the
+`perf`-based `count_instructions` shell function (see the section on `perf`) to
+judge whether a revision contains a bug looks something like this:
+
+```
+#!/bin/sh
+THRESHOLD=500000000
+CURR=$(git describe)
+utils/update-checkout --scheme master --reset-to-remote --clone --clean
+git checkout ${CURR}
+utils/update-checkout --scheme master --match-timestamp
+git checkout ${CURR}
+if utils/build-script -r
+then
+    V=$(count_instructions ../build/Ninja-ReleaseAssert/swift-linux-x86_64/bin/swiftc -c test.swift)
+    if [ ${V} -gt ${THRESHOLD} ]
+    then
+        # Bad
+        exit 1
+    else
+        # Good
+        exit 0
+    fi
+else
+    # Skip
+    exit 125
+fi
+```
+Note that in the example, the `utils/update-checkout` script is called twice,
+once to reset the adjacent repositories to their head state, and once with the
+`--match-timestamp` argument to match the adjacent repositories to the latest
+point in their history before the timestamp of the primary Swift repository
+being bisected. This mechanism is necessary if the regression range includes
+incompatible changes to `clang`, `llvm` or similar adjacent repositories.
+
+
+##### Creduce
+
+The `creduce` tool takes an input program and a script, and repeatedly cuts
+pieces out of the input program and re-runs the script on the program to see if
+the script still considers the residual input "interesting". It is
+documented [here](https://embed.cs.utah.edu/creduce/) and is available through
+Homebrew on macOS, or in most Linux package ecosystems.
+
+You can use `creduce` to automate the otherwise-tedious task of taking a failing
+or slow input and reducing it to "just the bad part" that triggers the problem
+you're after.
+
+For performance analysis, using `creduce` requires that the script can reliably
+tell when it's "still seeing a regression" and when it's not. This means having
+a reliable timer or cost-measurement tool; the simplest and most reliable we
+know of is `perf` so we'll assume you're using it here, via the
+`count_instructions` shell function described in the section on `perf`; but
+other measurement tools also work, for example using
+`utils/process-stats-dir.py` to test changes in performance counters. To use
```

Comment:
I suspect a lot of people will find it easiest to work on a Mac, maybe some examples that work there would be good too?

Reply:
I wish I had something command-line scriptable that worked simply-and-reliably-as-perf on a mac, but absent the counters we've embedded in the program itself, there doesn't seem to be one. The command-line interfaces to dtrace and instruments are all quite unreliable.

Reply:
I've had some luck using `dtrace` by increasing the buffer significantly.

Reply:
I was thinking something using the counters?